File: LeetCode-C++/Binary Tree Postorder Traversal.cpp
#define F(i, a, b) for(int i = a; i <= b; i++)
#define FB(i, a, b) for(int i = a; i >= b; i--)
#define MAX2(a, b) a>b?a:b
#define MIN2(a, b) a<b?a:b

#include <iostream>
#include <vector>
#include <queue>

using namespace std;

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    //typedef 
    typedef TreeNode* iterator;

    //methods
    void helper( iterator root, vector<int>& ret ){
        if( root->left != NULL ){
            helper(root->left, ret);
        }

        if( root->right != NULL ){
            helper(root->right, ret);
        }

        ret.push_back(root->val);

        return ;
    }

    vector<int > postorderTraversal(iterator root){
        vector<int> ret = vector<int>();

        if( root == NULL ){
            return ret;
        }

        helper( root, ret ); 

        return ret;
    }
};

File: LeetCode-Java/Binary Tree Postorder Traversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *	 int val;
 *	 TreeNode left;
 *	 TreeNode right;
 *	 TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
	//variables 

	//methods
	public static void postOrderHelper( TreeNode root, ArrayList<Integer> ret ){
		if( root.left != null ){
			postOrderHelper(root.left, ret);
		}

		if( root.right != null ){
			postOrderHelper(root.right, ret);
		}
		
		ret.add(root.val);

		return ; 
	}

	public ArrayList<Integer> postorderTraversal(TreeNode root) {
		ArrayList<Integer> ret = new ArrayList<Integer>();
		if( root == null ){
			return ret;
		}

		postOrderHelper( root, ret );

		return ret; 
	}
}
